### PR TITLE
Enforce envelope encryption at AEAD provider

### PR DESCRIPTION
**Pull request by yan**

### CODE REVIEW

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 1]
```diff
@@ -72,16 +75,24 @@ class CryptoModule(
...
-      return AeadFactory.getPrimitive(keysetHandle)
-          .also { keyManager[key.key_name] = it }
+      val kek = AeadFactory.getPrimitive(keysetHandle)
+      val envelopeKey = KmsEnvelopeAead(DEK_TEMPLATE, kek)
+
+      return envelopeKey.also { keyManager[key.key_name] = it }
     }
 
+    companion object {
+      val DEK_TEMPLATE : KeyTemplate = AeadKeyTemplates.AES128_GCM
```
Comment (mightyguava) [inline misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt]: You are not using this outside this file. Do you mean to use it in tests?

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 2]
```diff
@@ -72,16 +75,24 @@ class CryptoModule(
     }
   }
 
-  private class AeadProvider(val keyUri: String, val key: Key) : Provider<Aead> {
+  /**
+   * We do not support creating bare AEAD keys, only via envelope encryption.
```
Comment (mightyguava) [inline misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt]: nit: Probably better to word it as "we only support Aead keys via envelope encryption"

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt (hunk 1 above)
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt]: Just wanted to keep the DEK template visually separate from its use, if we want to adjust it later. Don't mind inlining it however.

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 3]
```diff
@@ -49,7 +52,7 @@ class CryptoModule(
         KeyType.AEAD -> {
           bind<Aead>()
               .annotatedWith(Names.named(key.key_name))
-              .toProvider(AeadProvider(config.kms_uri, key))
+              .toProvider(AeadEnvelopeProvider(config.kms_uri, key))
```
Comment (yoavamit) [inline misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt]: Might be worth adding documentation about this in the README file on in the `AeadKeyManager` javadoc

Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt]: 👍 updated the docstring to expand a bit on operating mode.

**Pull request merged**